haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -45,10 +57,37 @@ def _run_component(
             component_name=component_name, component=component, inputs=inputs
         )
 
+        # NOTE: a workaround for the DocumentJoiner and BranchJoiner components, since _consume_component_inputs()
+        # wraps 'documents' in an extra list, so if there's a 3 level deep list, we need to flatten it to 2 levels only
+        # ToDo: investigate why this is needed and if we can remove it
+        if self.resume_state and isinstance(instance, DocumentJoiner):  # noqa: SIM102
+            if isinstance(component_inputs["documents"], list):  # noqa: SIM102
+                if isinstance(component_inputs["documents"][0], list):  # noqa: SIM102
+                    if isinstance(component_inputs["documents"][0][0], list):  # noqa: SIM102
+                        component_inputs["documents"] = component_inputs["documents"][0]
+
+        if self.resume_state and isinstance(instance, BranchJoiner):  # noqa: SIM102
+            if isinstance(component_inputs["value"], list):  # noqa: SIM102
+                if isinstance(component_inputs["value"][0], list):  # noqa: SIM102
+                    if isinstance(component_inputs["value"][0][0], list):  # noqa: SIM102
+                        component_inputs["value"] = component_inputs["value"][0]
+
         # We need to add missing defaults using default values from input sockets because the run signature
         # might not provide these defaults for components with inputs defined dynamically upon component initialization
         component_inputs = self._add_missing_input_defaults(component_inputs, component["input_sockets"])
 
+        # Deserialize the inputs if they are passed in resume state
+        # this check will prevent other inputs generated at runtime from being deserialized
+        if self.resume_state and component_name in self.resume_state["pipeline_state"]["inputs"].keys():
+            for key, value in component_inputs.items():
+                component_inputs[key] = Pipeline._deserialize_component_input(value)
+
+        # add component_inputs to inputs
+        breakpoint_inputs = inputs.copy()
+        breakpoint_inputs[component_name] = Pipeline._remove_unserialisable_data(component_inputs)
```

Comment:
We're mixing serialize / serialise. For consistency, we should use _remove_unserializable_data